### PR TITLE
[SPARK-28916][SQL] Split subexpression elimination functions code for Generate[Mutable|Unsafe]Projection

### PR DESCRIPTION
**Pull request by mgaido91**
### What changes were proposed in this pull request?

The PR proposes to split the code for subexpression elimination before inlining the function calls all in the apply method for `Generate[Mutable|Unsafe]Projection`.

### Why are the changes needed?

Before this PR, code generation can fail due to the 64KB code size limit if a lot of subexpression elimination functions are generated. The added UT is a reproducer for the issue (thanks to the JIRA reporter and @HyukjinKwon for it).

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

added UT

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @cloud-fan @kiszk @maropu

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -2202,4 +2202,12 @@ class DataFrameSuite extends QueryTest with SharedSparkSession {
           |*(1) Range (0, 10, step=1, splits=2)""".stripMargin))
     }
   }
+
+  test("SPARK-28916: subexrepssion elimination can cause 64kb code limit") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: how long does this test run? We can write a unit test instead if the end-to-end test is too expensive to run.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -993,6 +994,14 @@ class CodegenContext {
     }
   }
 
+  /**
+   * Returns the code for subexpression elimination after splitting it if necessary.
+   */
+  def subexprFunctionsCode: String = {
+    // Wholestage codegen does not allow subexpression elimination
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: shall we add `assert(currentVars == null)` to guarantee it?

**Comment by cloud-fan** [comment]
@mgaido91 thanks for the fix! One question: AFAIK we will fallback to interpreted code path if codegen fails compilation. Why it doesn't work here？

**Comment by mgaido91** [comment]
> AFAIK we will fallback to interpreted code path if codegen fails compilation. Why it doesn't work here

It does work. We could see this as an improvement, indeed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: cc @maropu , I vaguely remember that we do subexpression elimination in hash aggregate in one of your PR.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: yes, it takes about 3 mins. I'll try and find a way to create a UT...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I added it, thanks. Anyway, when we generate subexpression elimination functions they all take an `InternalRow` as input. Unless that mechanism changes, I doubt it is usable in wholestage codegen.

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Could you elaborate on your thought on this comment?   
IMHO, this change is not directly related to whether wholestage codegen performs subexpression elimination by controlling in other places.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: The reason of this comment is that here we are passing the `InternalRow` to the variables and not the `currentVars` which would be needed in whole-stage codegen. This is done because in subexpression elimination we always put an `InternalRow` as argument. For this reason, whole-stage codegen disables subexpression elimination.
So I put here this comment in order to highlight that if in the future there will be a work to support subexpression elimination also in wholestage codegen, we need to modify this method too.

**Comment by mgaido91** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: > shall we add `assert(currentVars == null)` to guarantee it?

I cannot do that, it doesn't work. In that case just an empty string is returned because the subexpression functions seq is empty.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 3]
```diff
@@ -993,6 +994,15 @@ class CodegenContext {
     }
   }
 
+  /**
+   * Returns the code for subexpression elimination after splitting it if necessary.
+   */
+  def subexprFunctionsCode: String = {
+    // Wholestage codegen does not allow subexpression elimination: in that case, subexprFunctions
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I think this is not true, see `subexpressionEliminationForWholeStageCodegen`. @mgaido91 can you doule check to see if this fix works for whole-stage-codegen as well?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Seems like whole-stage-codegen has the same issue, but unfortunately `splitExpressionsWithCurrentInputs` is not completed yet.

I think the right description here is: whole-stage-codegen supports subexpression elimination, but we are not able to split the code for it yet.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: cc @maropu @viirya to double-check.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: @cloud-fan yes, sorry, this is not very well explained. I mean the wholestage codegen does not honor `spark.sql.subexpressionElimination.enabled` and even though it is true, it doesn't use `subexprFunctions`. So in this method we don't have to deal with wholestage codegen. Do you have suggestion on rewording this comment?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: how about `whole-stage-codegen supports subexpression elimination, and is handled by another code path`.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: thanks

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: currently looks better.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -2202,4 +2202,12 @@ class DataFrameSuite extends QueryTest with SharedSparkSession {
...
+    val df = spark.range(2).selectExpr((0 to 5000).map(i => s"id as field_$i"): _*)
+    df.createOrReplaceTempView("spark64kb")
+    val data = spark.sql("select * from spark64kb limit 10")
+    // This fails if 64Kb limit is reached in code generation
+    data.describe()
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Shall we disable fallback to interpreter in this test?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 5]
```diff
@@ -2202,4 +2202,12 @@ class DataFrameSuite extends QueryTest with SharedSparkSession {
           |*(1) Range (0, 10, step=1, splits=2)""".stripMargin))
     }
   }
+
+  test("SPARK-28916: subexrepssion elimination can cause 64kb code limit") {
+    val df = spark.range(2).selectExpr((0 to 5000).map(i => s"id as field_$i"): _*)
+    df.createOrReplaceTempView("spark64kb")
+    val data = spark.sql("select * from spark64kb limit 10")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Is this a test for `GenerateMutableProjection`? How about the case `GenerateUnsafeProjection`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 6]
```diff
@@ -993,6 +994,14 @@ class CodegenContext {
     }
   }
 
+  /**
+   * Returns the code for subexpression elimination after splitting it if necessary.
+   */
+  def subexprFunctionsCode: String = {
+    // Whole-stage codegen's subexpression elimination is handled in another code path
+    splitExpressions(subexprFunctions, "subexprFunc_split", Seq("InternalRow" -> INPUT_ROW))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: We don't need to check if its empty?
```
if (subexprFunctions.nonEmpty) {
  splitExpressions(...
} else {
 ""
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: looks ok, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: btw, (this is a off-topic though) the `HashAggregateExec` code for common subexpr elimination has the same issue? That also expands all generated the code for CSE in a single method now;
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L287

**Comment by maropu** [comment]
btw, (this is a off-topic though) the HashAggregateExec code for common subexpr elimination has the same issue? That also expands all generated the code for CSE in a single method now;
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L287

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: If so, how about `assert(currentVars != null && subexprFunctions.isEmpty)` for strict checks?

**Comment by maropu** [comment]
To make the title more precise, how about adding `For Generate[Mutable|Unsafe]Projection` in it?

**Comment by rednaxelafx** [comment]
The PR as is it now is one level better than the status quo. That's probably good enough. But I was curious whether or not it makes more sense to perform a tree-splitting instead of a fixed-level splitting.

Basically, `CodegenContext.splitExpressions` only performs a fixed one-level splitting, so it splits
```
orig_func() {
  expr1
  expr2
  expr3
  expr4
  expr5
  expr6
}
```
into something like the following, assuming the split threshold is an imaginary `2` expressions:
```
func1() { expr1; expr2 }
func2() { expe3; expr4 }
func3() { expr5; expr6 }

orig_func() {
  func1()
  func2()
  func3()
}
```
Now, given that we assume the split threshold is `2`, after the split this top-level code is still above the split threshold, which is not good.

Instead, it'd be really nice if the `splitExpressions` utility method can perform tree splitting within itself, without a 1- or 2-level fixed depth split limit.

Doing so would help us be more likely to cap the codegen method size below not only 64KB but also some lower thresholds like 8KB.

**Comment by kiszk** [comment]
@maropu good catch. In addition to that, [this line](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L291) may also cause a very huge method. I think that another PR can address these issues.

**Comment by kiszk** [comment]
@rednaxelafx I had a similar thought in my mind. I was optimistic that HotSpot compiler can apply inlining as possible.

It would be great if `splitExpressions` can handle splitting better.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 4 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: I need to revisit it according to Wenchen's suggestion... unfortunately this might take some time...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 6 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: It is not necessary: if it is empty `splitExpressions` would return an empty string, anyway I can add it if you think it is more clear

**Comment by mgaido91** [comment]
> btw, (this is a off-topic though) the HashAggregateExec code for common subexpr elimination has the same issue? That also expands all generated the code for CSE in a single method now;
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L287

@maropu I think the point there is that in case of wholestage codegen we are not able to split things at the moment, so I am not sure whether is is possible doing anything better there.

**Comment by cloud-fan** [comment]
@mgaido91 It will be possible after https://github.com/apache/spark/pull/20965 . We can apply this patch in `HashAggregateExec` later.

**Comment by mgaido91** [comment]
> But I was curious whether or not it makes more sense to perform a tree-splitting instead of a fixed-level splitting.

@rednaxelafx actually we are doing something similar to what you are suggesting. We already have a 2 or more levels splitting feature. The point is that the splitting point is given by the inner classes. We can say that now we assume that the number of function calls which fits in a single inner classes are a safe threshold for the number of function calls inside a specific function. This is not exactly what you are proposing, as it is not driven by the method size conf, but it is close. For more details please check `generateInnerClassesFunctionCalls` and SPARK-22226.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I can add that, thanks

**Comment by maropu** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!
